packages/bpk-component-infinite-scroll/src/BpkInfiniteScroll-test.js

```diff
@@ -0,0 +1,198 @@
+/*
+ * Backpack - Skyscanner's Design System
+ *
+ * Copyright 2018 Skyscanner Ltd
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React from 'react';
+import PropTypes from 'prop-types';
+import renderer from 'react-test-renderer';
+import { mount } from 'enzyme';
+import toJson from 'enzyme-to-json';
+import BpkInfiniteScroll from './BpkInfiniteScroll';
+import DEFAULT_REQUEST_ELEMENTS from './defaultRequestElements';
+
+global.IntersectionObserver = class {
+  observe() {} // eslint-disable-line
+  unobserve() {  } // eslint-disable-line
+};
+
+describe('BpkInfiniteScroll', () => {
+  const elementsArray = [];
+
+  for (let i = 0; i < 5; i += 1) {
+    elementsArray.push(`Element ${i}`);
+  }
+
+  const List = props => (
+    <div id="list">
+      {props.elements.map(element => <div key={element}>{element}</div>)}
+    </div>
+  );
+
+  List.propTypes = {
+    elements: PropTypes.arrayOf(PropTypes.string).isRequired,
+  };
+
+  const InfiniteList = BpkInfiniteScroll(List);
+
+  it('should render empty correctly', () => {
+    const tree = renderer.create(
+      <InfiniteList
+        onRequestElements={DEFAULT_REQUEST_ELEMENTS(elementsArray)}
+      />,
+    );
+
+    expect(tree).toMatchSnapshot();
+  });
+
+  it('should render non-empty correctly', async () => {
+    const tree = mount(
+      <InfiniteList
+        onRequestElements={DEFAULT_REQUEST_ELEMENTS(elementsArray)}
+      />,
+    );
+
+    const instance = tree.instance();
+    tree.setState(await instance.loadMore());
```

Comment:
I really don't like using `setState` during tests as this is supposed to be an internal api. What do you think we change this a bit to do a more "real world" test?

We can mock the intersectionApi:
```
let intersectionCallback = null;

global.IntersectionObserver = class {
  constructor(callback) {
    intersectionCallback = callback;
  }
  observe() {} // eslint-disable-line
  unobserve() {} // eslint-disable-line
};


describe('BpkInfiniteScroll', () => {
  afterEach(() => {
    intersectionCallback = null;
  });
```

The in the test we can do this:
```
it('should render non-empty correctly', async () => {
    const tree = renderer.create(
      <InfiniteList
        onRequestElements={DEFAULT_REQUEST_ELEMENTS(elementsArray)}
      />,
    );

    await intersectionCallback([{ intersectionRatio: 1 }]);
    expect(tree.toJSON()).toMatchSnapshot();
  });
```

We can do this for all the other tests.
For the above the work we need to change the `handleIntersection` a bit to return the promise. I'll add an inline comment explaining it.

Reply:
refactored the tests so we don't use `setState` anymore but an async `intersect` function that calls the IntersectionObserver callback.